exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileRecordReader.java

```diff
@@ -0,0 +1,146 @@
+
+import java.io.IOException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Stopwatch;
+import org.apache.drill.common.exceptions.DrillRuntimeException;
+import org.apache.drill.common.exceptions.ExecutionSetupException;
+import org.apache.drill.common.expression.SchemaPath;
+import org.apache.drill.common.types.TypeProtos;
+import org.apache.drill.common.types.Types;
+import org.apache.drill.exec.exception.SchemaChangeException;
+import org.apache.drill.exec.ops.OperatorContext;
+import org.apache.drill.exec.physical.impl.OutputMutator;
+import org.apache.drill.exec.record.MaterializedField;
+import org.apache.drill.exec.store.AbstractRecordReader;
+import org.apache.drill.common.types.TypeProtos.MajorType;
+import org.apache.drill.exec.vector.NullableVarBinaryVector;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.mapred.Reporter;
+import org.apache.hadoop.mapred.SequenceFileAsBinaryInputFormat;
+
+
+public class SequenceFileRecordReader extends AbstractRecordReader {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(SequenceFileRecordReader.class);
+
+  private static final int PER_BATCH_RECORD_COUNT = 4096;
+  private static final int PER_BATCH_BYTES = 256*1024;
+
+  private static final MajorType KEY_TYPE = Types.optional(TypeProtos.MinorType.VARBINARY);
+  private static final MajorType VALUE_TYPE = Types.optional(TypeProtos.MinorType.VARBINARY);
+
+  private final SchemaPath keySchema = SchemaPath.getSimplePath("binary_key");
+  private final SchemaPath valueSchema = SchemaPath.getSimplePath("binary_value");
+
+  private NullableVarBinaryVector keyVector;
+  private NullableVarBinaryVector valueVector;
+  private final FileSplit split;
+  private org.apache.hadoop.mapred.RecordReader<BytesWritable, BytesWritable> reader;
+  private final BytesWritable key = new BytesWritable();
+  private final BytesWritable value = new BytesWritable();
+  private final Configuration fsConf;
+
+  public SequenceFileRecordReader(final FileSplit split,
+                                  final Configuration fsConf) {
+    final List<SchemaPath> columns = new ArrayList();
+    columns.add(keySchema);
+    columns.add(valueSchema);
+    setColumns(columns);
+    this.fsConf = fsConf;
+    this.split = split;
+  }
+
+  @Override
+  protected boolean isSkipQuery() {
+    return false;
+  }
+
+  @Override
+  public void setup(OperatorContext context, OutputMutator output) throws ExecutionSetupException {
+    final SequenceFileAsBinaryInputFormat inputFormat = new SequenceFileAsBinaryInputFormat();
+    final JobConf jobConf = new JobConf(fsConf);
+    jobConf.setInputFormat(inputFormat.getClass());
+    try {
+      this.reader = inputFormat.getRecordReader(split, jobConf, Reporter.NULL);
```

Comment:
May be we should not defer this to the future then.

So for this to work, with impersonation enabled, the reader initialization needs to be done same as in [HiveRecordReader](https://github.com/apache/drill/pull/197/files#diff-ee3b9de759fc51b8b7fb908852c77e4dR232) using [runCallableAs](https://github.com/apache/drill/pull/197/files#diff-7254cd37acc418c43b47fa2ad3ddc6fcR58). The is because the record reader needs to be initialized within a proxy (for the file owner) user's context, otherwise a permission exception will be thrown.